MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java

```diff
@@ -2419,77 +2568,357 @@ private static void correctNonAeroFlyerBehavior(List<Entity> entityList, int boa
+                // that have nothing
+                if (hasGuided && tagCount > 0) {
+                    Arrays.fill(generatedBombs, BombType.B_NONE);
+                    generatedBombs[BombType.B_TAG]++;
+                    tagCount--;
+                    ((IBomber) curBomber).setBombChoices(generatedBombs);
                 }
 
-                loadBombs(entity, validBombChoices, campaign.getGameYear());
-                numBombers++;
             }
+
         }
+
     }
 
     /**
-     * Worker function that takes an entity and an array of bomb types
-     * and loads it up with as many of a mostly period-appropriate random bomb type
-     * as it's capable of holding
+     * Randomly generate a set of external ordnance up to the number of indicated bomb units. Lower
+     * rated forces are more likely to get simpler types (HE and rockets).
+     * @param bombUnits   how many bomb units to generate, some types count as more than one unit so
+     *                    returned counts may be lower than this but never higher
+     * @param airOnly     true to only select air-to-air ordnance
+     * @param isPirate    true if force is pirate, specific low-tech/high chaos selections
+     * @param quality     force rating to work with
+     * @param year        current year, for tech filter
+     * @return            array of integers, with each element being a bomb count using BombUnit
+     *                    enums as the lookup e.g. [BombUnit.HE] will get the number of HE
+     *                    bombs.
      */
-    private static void loadBombs(Entity entity, int[] validBombChoices, int year) {
-        int[] bombChoices = new int[BombType.B_NUM];
+    private static int[] generateExternalOrdnance (int bombUnits,
+                                                     boolean airOnly,
+                                                     boolean isPirate,
+                                                     int quality,
+                                                     int year) {
+
+        int[] bombLoad = new int[BombType.B_NUM];
+
+        if (bombUnits <= 0) {
+            return bombLoad;
+        }
+
+        // Get a random predefined loadout
+        String mapName = "";
+        double countWeight = 0.0;
+
+        Map<String, Integer> loadoutMap;
+        if (!isPirate) {
+            loadoutMap = bombMapGroundSpread;
+        } else {
+            loadoutMap = bombMapPirateGroundSpread;
+        }
+
+        double completeWeight = loadoutMap.values().stream().mapToDouble(curWeight -> Math.max(curWeight, 1.0)).sum();
+        double randomThreshold = (Compute.randomInt(100) / 100.0) * completeWeight;
+        for (String curMap : loadoutMap.keySet()) {
+            countWeight += Math.max(loadoutMap.get(curMap), 1.0);
+            if (countWeight >= randomThreshold) {
+                mapName = curMap;
+                break;
+            }
+        }
```

Comment:
This randomly chooses which entry loadout map type to apply to the unit by summing the relative weights in the loadout map, then checking a random value somewhere between 0.0 and (sum of weights) against a cumulative sum of the weights again?  And this works?

Just when I thought I couldn't hate arbitrary relative weighting any _more_...

My gut feeling is that this isn't going to work as planned, because bare Maps don't guarantee insertion-order ordering of entries.  So e.g. for `bombMapPirateGroundSpread`, a randomThreshold of <= 3.0 is equally likely to produce "Normal" or "Firestorm", and a value of 7.1+ will either select "Firestorm" or nothing; the chance of returning "Firestorm" is actually higher than the relative weighting of 7 and 3 would suggest.

Unfortunately I'm not an expert in this field so I'd need to go do some research to figure out the ideal approach, but I would say that switching the maps to LinkedHashMaps... unfortunately this requires using something other than Map.ofEntries() to populate them.  Perhaps a custom class that supports ofEntries()...

Reply:
Actually, looking at the code below that uses this selected string, I don't see why we couldn't just create array lists that give us weighted xBombLoad mappings, without the role names that are only used for the switch cases.

e.g., instead of `bombMapGroundSpread` plus a bunch of complicated weighted-option-selecting code _plus_ a switch on `mapName`, just cut out the middleman and do:

```
    public static ArrayList<Map<Integer, Integer>> createWeightedList(Map<Map<Integer, Integer>, Integer> weightMap) {
        // For each map M, there will be N entries where N is the arbitrary weight (integer)
        ArrayList<Map<Integer, Integer>> entriesByWeight = new ArrayList<Map<Integer, Integer>>();
        
        for(Map.Entry<Map<Integer, Integer>, Integer> m: weightMap.entrySet()) {
            // Create a sublist of size N = relative weight
            ArrayList<Map<Integer, Integer>> subList = new ArrayList<Map<Integer, Integer>>(m.getValue());
            for (int i=0;i<subList.size();i++) {
                // Fill with references to the specified Map of BombType:Weight
                subList.add(m.getKey());
            }
            // Add all entries
            entriesByWeight.addAll(subList);
        }
        return entriesByWeight;
    }
```

Pass it a Map of Maps and weights, like:
```
    private static final Map<String, Integer> standardBombMapsMap = Map.ofEntries (
            Map.entry(normalBombLoad, 6),
            Map.entry(antiMekBombLoad, 3),
            Map.entry(antiConvBombLoad, 2),
            Map.entry(standoffBombLoad, 1),
            Map.entry(strikeBombLoad, 2)
    );

    private static final ArrayList<Map<Integer, Integer>> standardBombWeightedList = 
        createWeightedList(standardBombMapsMap);
```

Then a randomized integer between 0 and the length of the resulting ArrayList can be used to randomly select one of the maps:

```
    Map<Integer, Integer> curMap = standardBombWeightedList.get(Compute.randomInt(standardBombWeightedList.size());
```

Which is all kind of verbose, but now you can just get the maps and the probability of getting a given map equals (weight / sum of weights).

Reply:
I've stepped through the code several dozen times during the development and testing process, and it does seem to produce reliable results at least as far as is intended.

As for the proposed substitute, that's getting very complicated, to the point where I'm having trouble following it. The (switch) is wordier and less elegant but also less opaque.

Don't want to put too much effort into fiddling with this, as eventually those distribution numbers are going to get pushed out to external data so players can fiddle with it to suit their needs.